src/VisualStudio/Core/Def/Implementation/UnusedReferences/RemoveUnusedReferencesCommandHandler.cs

```diff
@@ -0,0 +1,319 @@
+        }
+
+        private ImmutableArray<ReferenceUpdate> GetUnusedReferencesForProject(Project project, string projectAssetsFile, string targetFrameworkMoniker, CancellationToken cancellationToken)
+        {
+            ImmutableArray<ReferenceInfo> unusedReferences = ThreadHelper.JoinableTaskFactory.Run(async () =>
+            {
+                var projectReferences = await _referenceCleanupService.GetProjectReferencesAsync(project.FilePath!, cancellationToken).ConfigureAwait(true);
+                var references = ProjectAssetsReader.ReadReferences(projectReferences, projectAssetsFile, targetFrameworkMoniker);
+
+                return await _unusedReferencesService.GetUnusedReferencesAsync(project, references, cancellationToken).ConfigureAwait(true);
+            });
+
+            var referenceUpdates = unusedReferences
+                .Select(reference => new ReferenceUpdate(reference.TreatAsUsed ? UpdateAction.TreatAsUsed : UpdateAction.Remove, reference))
+                .ToImmutableArray();
+
+            return referenceUpdates;
+        }
+
+        private Window GetUnusedReferencesDialog(Project project, ImmutableArray<ReferenceUpdate> referenceUpdates)
+        {
+            var dialog = _unusedReferenceDialogProvider.CreateDialog(project, referenceUpdates);
+
+            var uiShell = _serviceProvider?.GetService<SVsUIShell, IVsUIShell>();
+            if (uiShell is null)
+            {
+                return dialog;
+            }
+
+            uiShell.GetDialogOwnerHwnd(out var ownerHwnd);
+
+            var windowHelper = new WindowInteropHelper(dialog)
+            {
+                Owner = ownerHwnd
+            };
+
+            uiShell.CenterDialogOnWindow(windowHelper.Handle, IntPtr.Zero);
+
+            return dialog;
+        }
+
+        private void ApplyUnusedReferenceUpdates(Project project, ImmutableArray<ReferenceUpdate> referenceUpdates, CancellationToken cancellationToken)
+        {
+            ThreadHelper.JoinableTaskFactory.Run(
+                () => _unusedReferencesService.UpdateReferencesAsync(project, referenceUpdates, cancellationToken));
+        }
+
+        private static bool TryGetPropertyValue(IVsHierarchy hierarchy, string propertyName, out string? propertyValue)
+        {
+            if (hierarchy is not IVsBuildPropertyStorage storage)
+            {
+                propertyValue = null;
+                return false;
+            }
+
+            return ErrorHandler.Succeeded(storage.GetPropertyValue(propertyName, null, (uint)_PersistStorageType.PST_PROJECT_FILE, out propertyValue));
+        }
+
+        private bool TryGetSelectedProjectHierarchy([NotNullWhen(returnValue: true)] out IVsHierarchy? hierarchy)
+        {
+            hierarchy = null;
+
+            // Get the DTE service and make sure there is an open solution
+            if (_serviceProvider?.GetService(typeof(EnvDTE.DTE)) is not EnvDTE.DTE dte ||
+                dte.Solution == null)
```

Comment:
This seems like an odd check...

Reply:
I stole it from the VisualStudioDiagnosticAnalyzerService and now it has moved into a shared set of helpers. Doesn't make it less odd tho.